cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -522,18 +522,49 @@ class UInt64Converter : public TypedConverterVisitor<UInt64Builder, UInt64Conver
 class DateConverter : public TypedConverterVisitor<Date64Builder, DateConverter> {
  public:
   inline Status AppendItem(const OwnedRef& item) {
-    auto pydate = reinterpret_cast<PyDateTime_Date*>(item.obj());
-    return typed_builder_->Append(PyDate_to_ms(pydate));
+    int64_t t;
+    if (PyDate_Check(item.obj())) {
+      auto pydate = reinterpret_cast<PyDateTime_Date*>(item.obj());
+      t = PyDate_to_ms(pydate);
+    } else {
+      t = static_cast<int64_t>(PyLong_AsLongLong(item.obj()));
```

Comment:
Seems this can fail: https://docs.python.org/2/c-api/long.html#c.PyLong_AsLongLong. Is there a way for some other kind of object than a PyLong to end up in this code branch?

Reply:
adding `RETURN_IF_PYERROR();` after this line is probably adequate error checking, then no further type checking is needed in the else

Reply:
fixed